### PR TITLE
[core] Hot reloading of integration configs

### PR DESCRIPTION
**Pull request by JohnLZeller**
This adds the ability to send a SIGUSR2 signal to the collector and successfully have it reload the `conf.d` integration configs.

Also note, this does not yet work on windows.

### CODE REVIEW

**Comment by JohnLZeller** [comment]
Ready for a first round of quick feedback @hkaj 
Still going to add the functions for the init.d script so it's callable. I will push those in a sec

File: packaging/debian/datadog-agent.init [hunk 1]
```diff
@@ -151,6 +152,11 @@ case "$1" in
         exit $?
         ;;
 
+    reload)
+        kill -12 `cat $COLLECTOR_PIDFILE`
```
Comment (remh) [inline packaging/debian/datadog-agent.init]: Watchout SIGUSR2 is not always 12, this is platform dependant. You should use `kill -SIGUSR2 instead`

File: checks/collector.py [hunk 2]
```diff
@@ -253,6 +253,8 @@ def run(self, checksd=None, start_event=True):
         """
         Collect data from each check and submit their data.
         """
+        log.info("Start of Collector.run()")
```
Comment (remh) [inline checks/collector.py]: We don't want these extra log lines at every iteration of the collection

Comment (JohnLZeller) [inline checks/collector.py]: I only left in logging to make it easier for reviewers to do testing, I will remove them in the final review.

File: packaging/debian/datadog-agent.init (hunk 1 above)
Comment (JohnLZeller) [inline packaging/debian/datadog-agent.init]: yeah, missed that :) Thanks

File: agent.py [hunk 3]
```diff
@@ -255,6 +292,7 @@ def main():
     if command in START_COMMANDS:
         log.info('Agent version %s' % get_version())
 
+    # MAYBE: Add reload command.. problem is that when called this way, agent.collector is None
```
Comment (LeoCavaille) [inline agent.py]: I think it's a no, we were trying to move away from these ad hoc commands, basically separate the management of the processes from the processes themselves.

File: agent.py [hunk 4]
```diff
@@ -131,6 +162,10 @@ def run(self, config=None):
 
         # Run the main loop.
         while self.run_forever:
+            log.info("Found {0} initialized checks".format(len(self._checksd['initialized_checks'])))
+            for che in self._checksd['initialized_checks']:
+                log.info("\t{0}".format(che))
```
Comment (LeoCavaille) [inline agent.py]: Logging this way is a bit of an anti-pattern, usually prefer:
```python
log.info("Found %d checks", len(checks))
```

File: agent.py [hunk 5]
```diff
@@ -78,9 +83,32 @@ def _handle_sigterm(self, signum, frame):
...
+        # Reload checksd configs
+        hostname = get_hostname(self._agentConfig)
+        self._checksd = load_check_directory(self._agentConfig, hostname)
+
+        # Logging
+        num_checks = len(self._checksd['initialized_checks'])
+        if num_checks > 0:
+            log.info("SIGUSR2: Successfully reloaded {num_checks} checks".
+                     format(num_checks=num_checks))
+            for che in self._checksd['initialized_checks']:
```
Comment (LeoCavaille) [inline agent.py]: nitpick: when there is no real need to shorten the variable names, there is no need :) always prefer clarity to brevity if the performance is the same

File: agent.py [hunk 6]
```diff
@@ -78,9 +83,32 @@ def _handle_sigterm(self, signum, frame):
         log.debug("Collector is stopped.")
 
     def _handle_sigusr1(self, signum, frame):
+        """Handles SIGUSR1, which signals an exit with an autorestart."""
         self._handle_sigterm(signum, frame)
         self._do_restart()
 
+    def _handle_sigusr2(self, signum, frame):
```
Comment (LeoCavaille) [inline agent.py]: I guess I would expect `SIGHUP` to implement such behavior in lieu of `SIGUSR2`.

**Comment by LeoCavaille** [comment]
Nice! Was not assigned on this one, but I was very excited to see this dream is finally real so I dropped a few remarks. :rocket:

**Comment by hkaj** [comment]
Great work @JohnLZeller ! Tested it on Ubuntu and CentOS, worked like a charm.
I don't have anything to add to the comments from Léo and Rémi except for the logging. Some `log.info` should be kept to ease troubleshooting in the future, but it should be short (no loop over the checks), yet helpful. Maybe a message when entering the signal handler, and another one just after reloading the configs.

File: packaging/debian/datadog-agent.init (hunk 1 above)
Comment (remh) [inline packaging/debian/datadog-agent.init]: as leo pointed out using SIGHUP has actually became a de facto standard, we should use it.

**Comment by JohnLZeller** [comment]
Thanks for the feedback everyone! :dancers:

What do we think about using `SIGUSR2` vs `SIGHUP`? I'm reluctant to using `SIGHUP` only because it seems more common than `SIGUSR2`.

File: packaging/centos/datadog-agent.init [hunk 7]
```diff
@@ -150,6 +151,14 @@ info() {
     exit $(($FORWARDER_RETURN+$COLLECTOR_RETURN+DOGSTATSD_RETURN))
 }
 
+reload() {
+    kill -HUP `cat $COLLECTOR_PIDFILE` 2>/dev/null
+    if [ $? -ne 0 ]; then
+        kill -SIGHUP `cat $COLLECTOR_PIDFILE`
```
Comment (hkaj) [inline packaging/centos/datadog-agent.init]: Not the cleanest code we have, but that will work on platforms that only have one or the other. :+1:

**Comment by JohnLZeller** [comment]
@hkaj Using `-1` instead of `-SIGHUP` or `-HUP`, because it seems that 1 is very consistently the HUP signal, and having to silence errors to try both is just too sloppy.

Lemme know what you think.

**Comment by hkaj** [comment]
@JohnLZeller actually we should use `-HUP`. We ship a `kill` command with the Linux agent, and it's based on `procps`, so `kill` will be consistent everywhere for the agent. Also mac os supports `-HUP` so we don't to make an exception there.
https://github.com/DataDog/dd-agent-omnibus/blob/master/config/projects/datadog-agent.rb#L105-L106

**Comment by JohnLZeller** [comment]
@hkaj changed to `HUP` and squashed commits. g2g :)

**Comment by hkaj** [comment]
Awesome, let's do it then!

**Comment by remh** [comment]
Watchout though, the kill command you call in the init script is not the one we ship unless you change the PATH.

**Comment by JohnLZeller** [comment]
Got it. Using the embedded kill commands in all scipts

**Comment by hkaj** [comment]
good catch @remh, lgtm now.

**Pull request merged**